<?xml version="1.0"?>
<robot name="inverted_pendulum" xmlns:xacro="http://ros.org/wiki/xacro">

<!-- this is a simplified version of inverted pendulum from inverted_pendulum_w_wheels
without wheels and with linear joint -->

<xacro:property name="M_PI" value="3.14159265" />

<xacro:property name="pendulum_length" value="0.3" />

<link name="world"/>

<link name="vehicle_body">
  <inertial>
    <mass value="1.0" />
    <inertia ixx="0.01" ixy="0.0" ixz="0.0"
      iyy="0.01" iyz="0.0"
      izz="0.01" />
  </inertial>

  <visual>
    <geometry>
      <box size="0.1 0.05 0.03"/>
    </geometry>
  </visual>

  <collision>
    <geometry>
      <box size="0.1 0.05 0.03"/>
    </geometry>
  </collision>
</link>

<gazebo reference="vehicle_body">
  <material>Gazebo/Grey</material>
</gazebo>

<joint name="vehicle_joint" type="prismatic">
  <parent link="world" />
  <child link="vehicle_body" />
  <origin xyz="0 0 0.02"/>
  <axis xyz="1 0 0" />
  <limit lower="-1" upper="1" effort="1000" velocity="1000" />
  <dynamics damping="0.0" />
</joint>

<!-- the inverted pendulum link -->
<link name="pendulum">
  <inertial>
    <mass value="0.1" />
    <origin xyz="0 0 ${pendulum_length/2}" />
    <inertia ixx="0.005" ixy="0.0" ixz="0.0"
      iyy="0.005" iyz="0.0"
      izz="0.001" />
  </inertial>

  <visual>
    <origin xyz="0 0 ${pendulum_length/2}" />
    <geometry>
      <cylinder radius="0.005" length="${pendulum_length}" />
    </geometry>
  </visual>

  <collision>
    <origin xyz="0 0 ${pendulum_length/2}" />
    <geometry>
      <cylinder radius="0.005" length="${pendulum_length}" />
    </geometry>
  </collision>
</link>

<gazebo reference="pendulum">
  <material>Gazebo/Blue</material>
</gazebo>

<!-- the inverted pendulum joint -->
<joint name="pendulum_joint" type="revolute">
  <parent link="vehicle_body" />
  <child link="pendulum" />
  <origin xyz="0 0 0.02"/>
  <axis xyz="0 1 0" />
  <limit lower="-${M_PI/2}" upper="${M_PI/2}" effort="1000" velocity="1000" />
  <!-- set limit for pendulum to not cross vehicle body -->
</joint>

<!-- the fixed mass at the top of the pendulum -->
<link name="pendulum_mass">
  <inertial>
    <mass value="1.0" />
    <inertia ixx="0.001" ixy="0.0" ixz="0.0"
      iyy="0.001" iyz="0.0"
      izz="0.001" />
  </inertial>

  <visual>
    <geometry>
      <sphere radius="0.01" />
    </geometry>
  </visual>

  <collision>
    <geometry>
      <sphere radius="0.01" />
    </geometry>
  </collision>
</link>

<gazebo reference="pendulum_mass">
  <material>Gazebo/Blue</material>
</gazebo>

<!-- fixed joint for pendulum mass -->
<joint name="pendulum_mass_joint" type="fixed">
  <parent link="pendulum" />
  <child link="pendulum_mass" />
  <origin xyz="0 0 ${pendulum_length}"/>
</joint>


</robot>

